### ISSUE TITLE
python-pptx fit text within text placeholder

### ISSUE THREAD

**Issue by soumyand** [scanny/python-pptx#715]
python-pptx while inserting text in predefined pptx template text placeholder i am using below options to fit text in text placeholder.
TP.auto_size = MSO_AUTO_SIZE.TEXT_TO_FIT_SHAPE
TP.word_wrap = True
TP.fit_text()

none of options fit text properly inside placeholder. text is overflowing out of placeholder. when opened ppt manually we have to set autofit text.

**Comment by Melendowski** [comment]
So out of curiosity what version of PowerPoint are you using?

Assuming the settings you gave above,

For >= PowerPoint 2016 when you open the font will not be autosized. However, simply saving the pptx file will cause the GUI to do the autosizing. 

For <= PowerPoint 2013, this autosizing never happesn, not even with a save. You can go into the settings of the placeholder and you'll see the correct settings but it wont trigger. What you need to do is de-select text-to-fit-shape and then re-select it, this will cause the sizing to occur.

**Comment by astrophysaxist** [comment]
I'm seeing a similar issue. Except after opening the created pptx file and saving the pptx, text remains overflowing. I'm using the latest powerpoint (Version 2107) which is through whatever enterprise account we have set up. I wonder if this a problem because this is a Microsoft 365 App.

**Comment by astrophysaxist** [comment]
Dropping a line to say I was able to get the autosizing to work in my pptx file by following similar instructions to the <=PowerPoint 2013 @Melendowski mentioned above. I opened the Slide Master, navigated to the slide layout in question; and on the placeholder shape's format menu Text Options tab, clicked 'Do not Autofit', then reselected 'Shrink text on overflow'. I exited the Slide Master and lo' and behold the text now fits in the placeholder box.

**Comment by MartinPacker** [comment]
That last might be automatable. Whether with AppleScript or VBA.

**Comment by scanny** [comment]
In my experience all it takes to trigger "re-flow" is to place the insertion-point in the text (click on text with mouse). At worst you need to add a space and then backspace over it. This is still a pain, but perhaps less likely to change some setting you didn't intend to. Interestingly, I've heard folks say this isn't a problem in LibreOffice, which automatically re-flows on load.

**Comment by MartinPacker** [comment]
I wonder if there's a VBA solution. But how to inject suitable VBA would be an interesting question. (As a Mac user I also wonder about AppleScript / JXA for this task.)

**Comment by MichaelPanwenjie** [comment]
i am still face the issue, anyone found the solution to set autofit automaticlly?

**Comment by shividhar** [comment]
Also facing the same issue

**Comment by wenchaoliu-93** [comment]
> Hello, thanks for the work for this package. @scanny
> 
> I'm really interested if someone found a solution to this problem meanwhile.
> 
> I'm using powerpoint 2016 and the text doesn't autofit. (textframe.autosize = MSO.FIT_TEXT_TO_SHAPE). I need to go open the presentation and add a whitespace and remove it so that it trigger the autofit like you mentionned all.
> 
> Anyone have a solution

I had the same issue. Yes, as mentioned above, there could be a solution via VBA. What helped was simply repositioning the textbox by dragging, which would trigger autofit.

**Comment by MuhammadSaboorIslam** [comment]
Facing the same issue, anyone found the solution ?

**Comment by kyleweise** [comment]
@scanny any updates here? I am facing a similar issue. Trying to use a combination of `MSO_AUTO_SIZE.TEXT_TO_FIT_SHAPE` and  `word_wrap = False` for auto-sized slide titles with no success. The documentation [here](https://python-pptx.readthedocs.io/en/latest/dev/analysis/shp-autofit.html) leads me to believe that this kind of functionality is not possible. 

Auto-sizing text to fit a shape seems like a _super_ common use-case for people using this package.

**Comment by Delengowski** [comment]
This is an issue of the program rendering. Nothing to do with python pptx and the xml is generates.

**Comment by MartinPacker** [comment]
Prompted to revisit this by @Delengowski 's comment...

... Recently I taught my md2pptx project to (re)open a presentation (and turn to a specific slide). You could - I suppose - reopen, navigate enough to cause rendering, and close again. That would force the Powerpoint engine to do its rendering. Then you could work on it again with python-pptx.

Today my project does the re-open and turn-to-a-slide using AppleScript. I talked to somebody recently who successfully experimented with PowerShell driving PowerPoint - so a Windows user could do it.

Quite how much it would take to render everything in a presentation I don't know.

(The intent of mentioning md2pptx, by the way, is not really to advertise it but to illustrate what I said could be done. md2pptx uses python-pptx to craft .pptx files.)

**Comment by Delengowski** [comment]
Injecting VBA script into the pptx file and it auto executing on load would solve this but of course that has a lot of bad side effects for security and Microsoft blocks it.

**Comment by MartinPacker** [comment]
Right. Which is why the "external orchestrator" route is what I suggested.
